plutus-ledger-api/src/Plutus/V1/Ledger/Api.hs

```diff
@@ -0,0 +1,196 @@
+    , Slot (..)
+    , SlotRange
+    -- *** Intervals
+    , Interval (..)
+    , Extended (..)
+    , Closure
+    , UpperBound (..)
+    , LowerBound (..)
+    -- *** Newtypes for script/datum types and hash types
+    , Validator (..)
+    , ValidatorHash (..)
+    , MonetaryPolicy (..)
+    , MonetaryPolicyHash (..)
+    , Redeemer (..)
+    , RedeemerHash (..)
+    , Datum (..)
+    , DatumHash (..)
+    -- * Errors
+    , EvaluationError (..)
+) where
+
+import           Control.Monad.Except
+import           Control.Monad.Writer
+import           Data.Aeson                                         (Result (..), fromJSON, toJSON)
+import           Data.Bifunctor
+import           Data.ByteString.Short
+import           Data.Either
+import           Data.Map
+import qualified Data.Text                                          as Text
+import           Data.Text.Prettyprint.Doc
+import           Data.Tuple
+import qualified Flat
+import qualified Language.PlutusCore                                as PLC
+import qualified Language.PlutusCore.Constant                       as PLC
+import qualified Language.PlutusCore.DeBruijn                       as PLC
+import           Language.PlutusCore.Evaluation.Machine.ExBudgeting (CostModel, ExBudget (..))
+import qualified Language.PlutusCore.Evaluation.Machine.ExBudgeting as PLC
+import           Language.PlutusCore.Evaluation.Machine.ExMemory    (ExCPU (..), ExMemory (..))
+import qualified Language.PlutusCore.MkPlc                          as PLC
+import           Language.PlutusCore.Pretty
+import           Language.PlutusTx                                  (Data (..), IsData (..))
+import qualified Language.PlutusTx.Lift                             as PlutusTx
+import qualified Language.UntypedPlutusCore                         as UPLC
+import qualified Language.UntypedPlutusCore.Evaluation.Machine.Cek  as UPLC
+import           Plutus.V1.Ledger.Address
+import           Plutus.V1.Ledger.Bytes
+import           Plutus.V1.Ledger.Contexts
+import           Plutus.V1.Ledger.Crypto
+import           Plutus.V1.Ledger.Interval
+import           Plutus.V1.Ledger.Scripts                           hiding (Script)
+import qualified Plutus.V1.Ledger.Scripts                           as Scripts
+import           Plutus.V1.Ledger.Slot
+
+{- Note [Abstract types in the ledger API]
+We need to support old versions of the ledger API as we update the code that it depends on. You
+might think that we should therefore make the types that we expose abstract, and only expose
+specific functions for constructing and working with them. However the situation is slightly
+different for us.
+
+Normally, when you are in this situation, you want to retain the same *interface* as the old version,
+but with the new types and functions underneath. Abstraction lets you do this easily. But we actually
+want to keep the old *implementation*, because things really have to work the same, bug-for-bug
+and byte-for-byte.
```

Comment:
Maybe clarify that byte-for-byte does *not* have anything to do with the script hashes, because this note is about the interpreter. That's why it shouldn't be a problem to build this module with a newer GHC for example. (Correct?)